aima-core/src/main/java/aima/core/nlp/ranking/HITS.java

```diff
@@ -0,0 +1,358 @@
+package aima.core.nlp.ranking;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Hashtable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Artificial Intelligence A Modern Approach (3rd Edition): page 871.<br>
+ * <br>
+ * 
+ * <pre>
+ * function HITS(query) returns pages (with hub and authority numbers)
+ *   pages <- EXPAND-PAGES(RELEVANT-PAGES(query))
+ *   for each p in pages do 
+ *   	p.AUTHORITY <- 1
+ *   	p.HUB <- 1
+ *   repeat until convergence do
+ *   	for each p in pages do
+ *   		p.AUTHORITY <- SUM<sub>i</sub> INLINK<sub>i</sub>(p).HUB
+ *   		p.HUB <- SUM<sub>i</sub> OUTLINK<sub>i</sub>(p).AUTHORITY
+ *   	NORMALIZE(pages)
+ *   return pages
+ * </pre>
+ * 
+ * Figure 22.1 The HITS algorithm for computing hubs and authorities with respect to a query.
+ * RELEVANT-PAGES fetches the pages that match the query, and EXPAND-PAGES add in every page
+ * that links to or is linked from one of the relevant pages. NORMALIZE divides each page's
+ * score by the sum of the squares of all pages' scores (separately for both the authority
+ * and hubs scores.<br>
+ * <br>
+ * 
+ * @author Jonathon Belotti (thundergolfer)
+ *
+ */
+public class HITS {
+
+	final int RANK_HISTORY_DEPTH = 3;
+	final double DELTA_TOLERANCE = 0.05; // somewhat arbitrary
+	Page[] pagesIndex;
+	Hashtable<String,Page> pagesTable;
+	// DETECT CONVERGENCE VARS
+    public double[] prevAuthVals;
+    public double[] prevHubVals;
+	public double prevAveHubDelta = 0;
+	public double prevAveAuthDelta = 0; 
+	////////////////////////////
+    
+    // TODO: Improve the convergence detection functionality
+    
+    public HITS( Hashtable<String,Page> pTable ) {
+    	this.pagesTable = pTable;
+    }
+	
+	public List<Page> rank( String query ) { 
+		List<Page> pages;
+		pages = expandPages( relevantPages( query, pagesTable ), pagesTable );
+		// for each p in pages
+		for( int i=0; i < pages.size(); i++ ) { 
+			Page p = pages.get(i);
+			p.authority = 1;
+			p.hub = 1; 
+		}
+		while( !convergence( pages ) ) {
+			// for each p in pages do
+			for( int i=0; i < pages.size(); i++ ) {
+				Page p = pages.get(i);
+				p.authority = SumInlinkHubScore( p );
+				p.hub       = SumOutlinkAuthorityScore( p );
+			}
+			normalize(pages);
+			try {
+				TimeUnit.SECONDS.sleep(1);
+			} catch (InterruptedException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+			}
+		}
+		return pages;
+		
+	}
+	
+	/**
+	 * Fetches and returns all pages that match the query
+	 * @param query
+	 * @return
+	 * @throws UnsupportedEncodingException 
+	 */
+	public List<Page> relevantPages( String query, Hashtable<String,Page> pTable ) {
+		List<Page> relevantPages = new ArrayList<Page>();
+		Iterator<String> it = pTable.keySet().iterator();
+		while(it.hasNext()) {
+			Page p = pTable.get(it.next());
```

Comment:
A more concise version of this is just a for each on the values, i.e.:

for (Page p : pTable.values()) {
....